backends/ubpf/frontend.cpp

```diff
@@ -0,0 +1,166 @@
+#include "ubpfModel.h"
+#include "frontend.h"
+#include "frontends/p4/frontend.h"
+#include "frontends/p4/typeMap.h"
+#include "frontends/common/resolveReferences/resolveReferences.h"
+#include "frontends/p4/typeChecking/bindVariables.h"
+#include "ir/pass_manager.h"
+
+// Passes
+#include "frontends/p4/actionsInlining.h"
+#include "frontends/p4/checkConstants.h"
+#include "frontends/p4/checkNamedArgs.h"
+#include "frontends/p4/createBuiltins.h"
+#include "frontends/p4/defaultArguments.h"
+#include "frontends/p4/deprecated.h"
+#include "frontends/p4/directCalls.h"
+#include "frontends/p4/dontcareArgs.h"
+#include "frontends/p4/evaluator/evaluator.h"
+#include "frontends/common/constantFolding.h"
+#include "frontends/p4/functionsInlining.h"
+#include "frontends/p4/hierarchicalNames.h"
+#include "frontends/p4/inlining.h"
+#include "frontends/p4/localizeActions.h"
+#include "frontends/p4/moveConstructors.h"
+#include "frontends/p4/moveDeclarations.h"
+#include "frontends/p4/parserControlFlow.h"
+#include "frontends/p4/removeReturns.h"
+#include "frontends/p4/resetHeaders.h"
+#include "frontends/p4/setHeaders.h"
+#include "frontends/p4/sideEffects.h"
+#include "frontends/p4/simplify.h"
+#include "frontends/p4/simplifyDefUse.h"
+#include "frontends/p4/simplifyParsers.h"
+#include "frontends/p4/specialize.h"
+#include "frontends/p4/specializeGenericFunctions.h"
+#include "frontends/p4/strengthReduction.h"
+#include "frontends/p4/structInitializers.h"
+#include "frontends/p4/switchAddDefault.h"
+#include "frontends/p4/tableKeyNames.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/p4/uniqueNames.h"
+#include "frontends/p4/unusedDeclarations.h"
+#include "frontends/p4/uselessCasts.h"
+#include "frontends/p4/validateMatchAnnotations.h"
+#include "frontends/p4/validateParsedProgram.h"
+
+namespace UBPF {
+
+class getUBPFModelVersion : public Inspector {
+    bool preorder(const IR::Declaration_Constant *dc) override {
+        if (dc->name == "__ubpf_model_version") {
+            auto val = dc->initializer->to<IR::Constant>();
+            UBPFModel::instance.version = static_cast<unsigned>(val->value); }
+        return false; }
+    bool preorder(const IR::Declaration *) override { return false; }
+};
+
+const IR::P4Program *FrontEnd::run(const CompilerOptions &options, const IR::P4Program *program,
+                                   bool skipSideEffectOrdering, std::ostream *outStream) {
+    if (program == nullptr && options.listFrontendPasses == 0)
+        return nullptr;
+
+    bool isv1 = options.isv1();
+    P4::ReferenceMap refMap;
+    P4::TypeMap typeMap;
+    refMap.setIsV1(isv1);
+
+    auto evaluator = new P4::EvaluatorPass(&refMap, &typeMap);
+    std::initializer_list<Visitor *> frontendPasses = {
```

Comment:
If this list is the same as in the standard front-end, then I would rather suggest to insert your pass in front, then call the standard front-end unmodified. This will make maintenance much simpler.

Reply:
Done